### ISSUE TITLE
Bug in the CriticalLines class.

### ISSUE THREAD

**Issue by IanMarteens** [sweisser/critical_lines_java#1]
It's a small one:

```
            alphav[j] = alpha;
            betav[j] = beta;

            if (j < inputVars.getNumVariables() - 1) { }
```

It should be ``j < inputVars.getNumVariables()``, without decrementing by 1. I've just found it because I had some test that produced negative weights.

Thanks by your code, by the way.

**Comment by sweisser** [comment]
How did your input variables look like for that test?

**Comment by IanMarteens** [comment]
Nothing special. It just happened than in one of the steps the variable going in was the variable with the highest index. I can send you the data if you want it. 11 securities, upper and lower bounds and just the full investment constraint.

**Comment by sweisser** [comment]
Yes, that would be helpful to reproduce and check it against the original (VBA) code.

**Comment by IanMarteens** [comment]
Sure. Here it is:

```
    public void read1(InputVariables inputVars) {
        inputVars.setEndLambdaE(0.00001);

        char[] conTypes = {'='};
        inputVars.init(11, conTypes);

        //
        // Constraints
        //
        double[][] tmpA =
        {
            { 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00, 1.00 },
        };
        double[] tmpB = { 1.0 };
        inputVars.setConstraints(tmpA, tmpB);

        //
        // Lower and upper limits, expected returns.
        //
        double[] tmpL = { 0.003, 0.000, 0.000, 0.000, 0.000, 0.000, 0.000, 0.000, 0.000, 0.002, 0.002 };
        inputVars.setLowerBoundaries(tmpL);

        double[] tmpU = { 0.003, 0.350, 0.350, 0.350, 0.100, 0.100, 0.350, 0.350, 0.350, 0.100, 0.100 };
        inputVars.setUpperBoundaries(tmpU);

        double[] tmpMu =
        {
            -0.00180453535071047, -0.00159576562379307, 0.00504580504430248, 0.00693673030759708, 0.02025457623320640, 0.01881327391737990, 0.00819210032583716, 0.04959567136167170, 0.04883020747769260, 0.04458828189108920, 0.05638728131532670,
        };
        inputVars.setExpectedReturns(tmpMu);

        double[][] tmpCov =
        {
            { 1.75032752535272E-07,  4.16815790023598E-06,  1.09013327353007E-05,  6.46384607479174E-06, 1.02465696051639E-05,    1.05416967482231E-05,    1.50054738205233E-05,    1.14779527601482E-05,    1.83172440963750E-06,    -1.65290582646624E-06,   1.00611245335712E-05, },
            { 4.16815790023598E-06,  1.35424347501119E-05,  3.94346735773140E-05,  1.58353715471880E-05, 8.31825121087282E-06,    1.84479755409853E-05,    4.00111483491390E-05,    1.42975096463588E-05,    -3.13228419428149E-06,   7.55083453713765E-06,    5.71189853580017E-06, },
            { 1.09013327353007E-05,  3.94346735773140E-05,  2.03077813068697E-04,  8.28038391861013E-05, 3.51687229590016E-06,    3.04146373101172E-05,    1.76633342868116E-04,    8.71942590644389E-06,    -8.73672589729711E-06,   9.66584807065849E-06,    6.13511689721174E-06, },
            { 6.46384607479174E-06,  1.58353715471880E-05,  8.28038391861013E-05,  7.83599371352068E-05, 9.08306034441343E-05,    8.85415854124304E-05,    1.00366676992752E-04,    1.36205169583194E-04,    9.53899497458210E-05,    8.69295728533315E-05,    1.57411258984934E-04, },
            { 1.02465696051639E-05,  8.31825121087282E-06,  3.51687229590016E-06,  9.08306034441343E-05, 3.82908262861540E-04,    3.76408518530427E-04,    9.73008916892730E-05,    4.80235018266035E-04,    2.24179437437307E-04,    1.35500814068050E-04,    4.90671630002765E-04, },
            { 1.05416967482231E-05,  1.84479755409853E-05,  3.04146373101172E-05,  8.85415854124304E-05, 3.76408518530427E-04,    5.90090134502874E-04,    1.19053723649872E-04,    4.30077981657510E-04,    9.11566011847697E-05,    3.01122228826443E-05,    5.20442297877776E-04, },
            { 1.50054738205233E-05,  4.00111483491390E-05,  1.76633342868116E-04,  1.00366676992752E-04, 9.73008916892730E-05,    1.19053723649872E-04,    2.31034562313384E-04,    1.86684778887628E-04,    9.67150813869892E-05,    1.19214319511525E-04,    1.69497573736519E-04, },
            { 1.14779527601482E-05,  1.42975096463588E-05,  8.71942590644389E-06,  1.36205169583194E-04, 4.80235018266035E-04,    4.30077981657510E-04,    1.86684778887628E-04,    1.25635399730801E-03,    8.64638952112010E-04,    7.63842660936492E-04,    9.43065202860434E-04, },
            { 1.83172440963750E-06,  -3.13228419428149E-06, -8.73672589729711E-06, 9.53899497458210E-05, 2.24179437437307E-04,    9.11566011847697E-05,    9.67150813869892E-05,    8.64638952112010E-04,    1.15295268921757E-03,    8.90587116868042E-04,    8.06621421335402E-04, },
            { -1.65290582646624E-06, 7.55083453713765E-06,  9.66584807065849E-06,  8.69295728533315E-05, 1.35500814068050E-04,    3.01122228826443E-05,    1.19214319511525E-04,    7.63842660936492E-04,    8.90587116868042E-04,    1.49544134549717E-03,    7.48919958892219E-04, },
            { 1.00611245335712E-05,  5.71189853580017E-06,  6.13511689721174E-06,  1.57411258984934E-04, 4.90671630002765E-04,    5.20442297877776E-04,    1.69497573736519E-04,    9.43065202860434E-04,    8.06621421335402E-04,    7.48919958892219E-04,    1.54404845021622E-03, },
        };

        inputVars.setCovarianceMatrix(tmpCov);
    }
```

**Comment by IanMarteens** [comment]
The last two portfolios from the frontiers have a couple of negative weights.

**Comment by sweisser** [comment]
Yes, that is wrong given the lower limits... With the change you suggested, it works fine.
I put it into the VBA code and Excel that I typed in from the original book, but it get errors there and even different results than the Java implementation. Guess the whole code needs a few (JUnit) test cases and a lot of refactoring. Currently its basically Java in the structure of the old VBA stuff.

I checked in your proposed solution along with the example.
